### PR TITLE
[Internal] Variable storage

### LINKED ISSUE #16129: [Test] Add Variable Storage for random names in tests
Link to test proxy documentation: https://github.com/Azure/azure-sdk-tools/tree/main/tools/test-proxy/Azure.Sdk.Tools.TestProxy#storing-variables

This is useful for random items (tables, keys, messages) that can be stored in the recordings and delivered on test set up.

### PR DESCRIPTION
**Pull request by seankane-msft**
Closes #16129 

Adds a variable storage for storing random names and retrieving them in future tests.

### CODE REVIEW

File: sdk/internal/recording/recording.go [hunk 1]
```diff
@@ -585,11 +588,30 @@ func Start(t *testing.T, pathToRecordings string, options *RecordingOptions) err
 		}
 		return fmt.Errorf("Recording ID was not returned by the response. Response body: %s", b)
 	}
+
+	// Unmarshal any variables returned by the proxy
+	var m map[string]interface{}
+	body, err := ioutil.ReadAll(resp.Body)
```
Comment (jhendrixMSFT) [inline sdk/internal/recording/recording.go:594]: Looks like `resp.Body` isn't being closed.  We always need to call `resp.Body.Close()`.

File: sdk/internal/recording/recording.go [hunk 2]
```diff
@@ -614,13 +636,30 @@ func Stop(t *testing.T, options *RecordingOptions) error {
 	if err != nil {
 		return err
 	}
+	if options.Variables != nil && len(options.Variables) > 0 {
```
Comment (jhendrixMSFT) [inline sdk/internal/recording/recording.go:639]: You can omit the `nil` check.  The length of a nil map/slice is always zero.

File: sdk/internal/recording/recording.go [hunk 3]
```diff
@@ -614,13 +636,30 @@ func Stop(t *testing.T, options *RecordingOptions) error {
...
+	}
+
 	var recTest recordedTest
 	var ok bool
 	if recTest, ok = testSuite[t.Name()]; !ok {
 		return errors.New("Recording ID was never set. Did you call StartRecording?")
 	}
 	req.Header.Set("x-recording-id", recTest.recordingId)
-	_, err = client.Do(req)
+	resp, err := client.Do(req)
```
Comment (jhendrixMSFT) [inline sdk/internal/recording/recording.go:655]: `defer resp.Body.Close()`

File: sdk/internal/recording/recording.go [hunk 4]
```diff
@@ -721,3 +760,10 @@ func IsLiveOnly(t *testing.T) bool {
 	}
 	return false
 }
+
+func GetVariables(t *testing.T) map[string]interface{} {
```
Comment (jhendrixMSFT) [inline sdk/internal/recording/recording.go:764]: Doc comment

**Review by jhendrixMSFT — APPROVED** [review]

**Comment by seankane-msft** [comment]
After #16355  is merged, I will prepare this for release and merge.

**Pull request merged**